File: src/util/hotkey/Hotkeys.java
package util.hotkey;

import java.lang.reflect.Field;
import java.util.List;
import java.util.Map;
import java.util.concurrent.AbstractExecutorService;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import javafx.scene.input.KeyCode;
import org.jnativehook.GlobalScreen;
import org.jnativehook.NativeHookException;
import org.jnativehook.NativeInputEvent;
import org.jnativehook.keyboard.NativeKeyEvent;
import org.jnativehook.keyboard.NativeKeyListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import util.Util;
import util.action.Action;
import static util.functional.Util.list;
import static util.functional.Util.stream;

/**
 * Global hotkey manager, implemented on top of https://github.com/kwhat/jnativehook library
 */
public class Hotkeys {

	private static final Logger LOGGER = LoggerFactory.getLogger(Hotkeys.class);

	private final Consumer<Runnable> executor;
	private final Map<Integer,KeyCombo> keyCombos = new ConcurrentHashMap<>();
	private NativeKeyListener keyListener;

	public Hotkeys(Consumer<Runnable> executor) {
		this.executor = executor;
	}

	public void start() {
		if (!isRunning()) {
			// Disable library logging.
			java.util.logging.Logger logger = java.util.logging.Logger.getLogger(GlobalScreen.class.getPackage().getName());
			logger.setLevel(java.util.logging.Level.OFF);
			logger.setUseParentHandlers(false);

			try {
				GlobalScreen.setEventDispatcher(new AbstractExecutorService() {
					private boolean running = true;

					@Override
					public void shutdown() {
						this.running = false;
					}

					@Override
					public List<Runnable> shutdownNow() {
						this.running = false;
						return list();
					}

					@Override
					public boolean isShutdown() {
						return !this.running;
					}

					@Override
					public boolean isTerminated() {
						return !this.running;
					}

					@Override
					public boolean awaitTermination(long amount, TimeUnit units) throws InterruptedException {
						return true;
					}

					@Override
					public void execute(Runnable action) {
//						executor.accept(action);
						action.run();
					}
				});
				GlobalScreen.registerNativeHook();
				keyListener = new NativeKeyListener() {
					@Override
					public void nativeKeyPressed(NativeKeyEvent e) {
						// LOGGER.info("Global key press event: " + e.paramString());

						keyCombos.forEach((actionId, keyCombo) -> {
							Action a = Action.get(actionId);

							// TODO: remove hack
							// For some reason left BACK_SLASH key (left of the Z key) is not recognized, so we manually se it straight
							if (e.getRawCode()==226) {
								e.setKeyCode(43);
								e.setKeyChar((char) 43);
							}

							// Unfortunately, JavaFX key codes and the library raw codes do not match for some keys, so we also
							// check key name. This combination should be enough for all but rare cases
							boolean modifierMatches = keyCombo.modifier==e.getModifiers();
							boolean keysMatch = keyCombo.key.getCode()==e.getRawCode() ||
									keyCombo.key.getName().equalsIgnoreCase(NativeKeyEvent.getKeyText(e.getKeyCode()));
							if (keysMatch && modifierMatches)
								keyCombo.press(a, e);
						});
					}

					@Override
					public void nativeKeyReleased(NativeKeyEvent e) {
						keyCombos.values().stream().filter(k -> k.isPressed).forEach(k -> k.release(e));
					}

					@Override
					public void nativeKeyTyped(NativeKeyEvent e) {}
				};
				GlobalScreen.addNativeKeyListener(keyListener);
			} catch (NativeHookException e) {
				LOGGER.error("Failed to register global hotkeys", e);
			}
		}
	}

	public boolean isRunning() {
		return GlobalScreen.isNativeHookRegistered();
	}

	public void stop() {
		if (isRunning()) {
			try {
				GlobalScreen.removeNativeKeyListener(keyListener);
				keyListener = null;
				GlobalScreen.unregisterNativeHook();
			} catch (NativeHookException e) {
				LOGGER.error("Failed to unregister global hotkeys", e);
			}
		}
	}

	public void register(Action action, String keys) {
		int i = keys.lastIndexOf('+');
		String keyString = (i<0 ? keys : keys.substring(i + 1)).trim().toLowerCase().replace('_', ' ');
		KeyCode key = stream(KeyCode.values()).findFirst(k -> k.getName().toLowerCase().equals(keyString)).get();
		register(
				action, key,
				stream(KeyCode.ALT, KeyCode.ALT_GRAPH, KeyCode.SHIFT, KeyCode.CONTROL, KeyCode.WINDOWS, KeyCode.COMMAND, KeyCode.META)
						.filter(k -> Util.containsNoCase(keys, k.name()) || (k==KeyCode.CONTROL && keys.toUpperCase().contains("CTRL")))
						.toArray(KeyCode[]::new)
		);
	}

	public void register(Action action, KeyCode key, KeyCode... modifiers) {
		keyCombos.put(action.getID(), new KeyCombo(key, modifiers));
	}

	public void unregister(Action action) {
		keyCombos.remove(action.getID());
	}

	private final class KeyCombo {
		public final KeyCode key;
		public final KeyCode[] modifiers;
		public final int modifier;
		private boolean isPressed = false;

		public KeyCombo(KeyCode key, KeyCode[] modifiers) {
			this.key = key;
			this.modifiers = modifiers;

			int m = 0;
			if (stream(modifiers).has(KeyCode.SHIFT)) m += 1;   // right shift
			if (stream(modifiers).has(KeyCode.CONTROL)) m += 2; // left ctrl
			if (stream(modifiers).has(KeyCode.WINDOWS)) m += 4;
			if (stream(modifiers).has(KeyCode.ALT)) m += 8;
//			if(stream(modifiers).has(KeyCode.SHIFT)) m += 16;   // right shift
//			if(stream(modifiers).has(KeyCode.CONTROL)) m += 32; // right ctrl
			if (stream(modifiers).has(KeyCode.COMMAND)) m += 64;
			if (stream(modifiers).has(KeyCode.ALT_GRAPH)) m += 128;
//			if(stream(modifiers).has(KeyCode.META)) m += 128;
			this.modifier = m;
		}

		public void press(Action a, NativeKeyEvent e) {
			boolean isPressedFirst = !isPressed;
			isPressed = true;
			if (a.isContinuous() || isPressedFirst) {

				// consume event (must be on the jNativeHook thread)
				try {
					Field f = NativeInputEvent.class.getDeclaredField("reserved");
					f.setAccessible(true);
					f.setShort(e, (short) 0x01);
				} catch (Exception x) {
					LOGGER.error("Failed to consume native key event.", e);
				}

				executor.accept(a);
			}
		}

		@SuppressWarnings("unused")
		synchronized public void release(NativeKeyEvent e) {
			isPressed = false;
		}
	}
}

File: src/sp/it/pl/util/hotkey/Hotkeys.kt
package sp.it.pl.util.hotkey

import javafx.scene.input.KeyCode
import mu.KLogging
import org.jnativehook.GlobalScreen
import org.jnativehook.NativeHookException
import org.jnativehook.NativeInputEvent
import org.jnativehook.keyboard.NativeKeyEvent
import org.jnativehook.keyboard.NativeKeyListener
import sp.it.pl.util.action.Action
import sp.it.pl.util.dev.fail
import sp.it.pl.util.functional.Util.list
import java.util.concurrent.AbstractExecutorService
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/** Global hotkey manager, implemented on top of https://github.com/kwhat/jnativehook library. */
class Hotkeys {
    private val executor: (Runnable) -> Unit
    private val keyCombos = ConcurrentHashMap<Int, KeyCombo>()
    private var keyListener: NativeKeyListener? = null
    private var isRunning = false

    constructor(executor: (Runnable) -> Unit) {
        this.executor = executor

        // Disable library logging.
         java.util.logging.Logger.getLogger(GlobalScreen::class.java.getPackage().name).apply {
             level = java.util.logging.Level.OFF
             useParentHandlers = false
         }
    }

    fun isRunning(): Boolean = isRunning

    fun start() {
        if (!isRunning) {
            logger.info { "Starting global hotkeys" }
            isRunning = true

            val eventDispatcher = object: AbstractExecutorService() {
                private var running = true

                override fun shutdown() {
                    running = false
                }

                override fun shutdownNow(): List<Runnable> {
                    running = false
                    return list()
                }

                override fun isShutdown() = !running
                override fun isTerminated() = !running
                override fun awaitTermination(amount: Long, units: TimeUnit) = true
                override fun execute(action: Runnable) = action.run()
            }
            val keyListener = object: NativeKeyListener {
                override fun nativeKeyPressed(e: NativeKeyEvent) {
                    keyCombos.forEach { actionId, keyCombo ->
                        val action = Action.get(actionId)

                        // For some reason left BACK_SLASH key (left of the Z key) is not recognized, recognize manually
                        if (e.rawCode==226) {
                            e.keyCode = 43
                            e.keyChar = 43.toChar()
                        }

                        // Unfortunately, JavaFX key codes and the library raw codes do not match for some keys, so we also
                        // check key name. This combination should be enough for all but rare cases
                        val modifiersMatch = keyCombo.modifier==e.modifiers
                        val keysMatch = keyCombo.key.code==e.rawCode || keyCombo.key.getName().equals(NativeKeyEvent.getKeyText(e.keyCode), ignoreCase = true)
                        if (keysMatch && modifiersMatch)
                            keyCombo.press(action, e)
                    }
                }

                override fun nativeKeyReleased(e: NativeKeyEvent) {
                    keyCombos.values.forEach { if (it.isPressed) it.release() }
                }

                override fun nativeKeyTyped(e: NativeKeyEvent) {}
            }
            try {
                GlobalScreen.setEventDispatcher(eventDispatcher)
                GlobalScreen.addNativeKeyListener(keyListener)
                GlobalScreen.registerNativeHook()
                this.keyListener = keyListener
            } catch (e: NativeHookException) {
                logger.error("Failed to register global hotkeys", e)
            }

        }
    }

    fun stop() {
        if (isRunning) {
            try {
                logger.info { "Stopping global hotkeys" }
                GlobalScreen.removeNativeKeyListener(keyListener)
                keyListener = null
                GlobalScreen.unregisterNativeHook()
            } catch (e: NativeHookException) {
                logger.error(e) { "Failed to unregister global hotkeys" }
            }
            isRunning = false
        }
    }

    fun register(action: Action, keys: String) {
        val keyString = keys.substringAfterLast('+').trim().toLowerCase().replace('_', ' ')
        val key = KeyCode.values().find { k -> k.getName().equals(keyString, true) }
                ?: fail { "No KeyCode for ${action.keys}" }
        register(
                action,
                key,
                *sequenceOf(KeyCode.ALT, KeyCode.ALT_GRAPH, KeyCode.SHIFT, KeyCode.CONTROL, KeyCode.WINDOWS, KeyCode.COMMAND, KeyCode.META)
                        .filter { k -> keys.contains(k.name, true) || (k==KeyCode.CONTROL && keys.contains("CTRL", true)) }
                        .toList().toTypedArray()
        )
    }

    fun register(action: Action, key: KeyCode, vararg modifiers: KeyCode) {
        keyCombos[action.id] = KeyCombo(key, *modifiers)
    }

    fun unregister(action: Action) {
        keyCombos.remove(action.id)
    }

    private inner class KeyCombo {
        val key: KeyCode
        val modifier: Int
        var isPressed = false
            private set(value) {
                field = value
            }

        constructor(key: KeyCode, vararg modifiers: KeyCode) {
            this.key = key
            this.modifier = run {
                var m = 0
                infix fun KeyCode.toMask(mask: Int) {
                    if (this in modifiers)
                        m += mask
                }
                KeyCode.SHIFT toMask NativeInputEvent.SHIFT_L_MASK
                KeyCode.CONTROL toMask NativeInputEvent.CTRL_L_MASK
                KeyCode.ALT toMask NativeInputEvent.ALT_L_MASK
                KeyCode.COMMAND toMask NativeInputEvent.META_R_MASK
                KeyCode.ALT_GRAPH toMask NativeInputEvent.ALT_R_MASK
                KeyCode.WINDOWS toMask NativeInputEvent.META_L_MASK
                KeyCode.META toMask NativeInputEvent.META_L_MASK
                m
            }
        }

        fun press(a: Action, e: NativeKeyEvent) {
            val isPressedFirst = !isPressed
            isPressed = true
            if (a.isContinuous || isPressedFirst) {
                // consume event (must be on the jNativeHook thread)
                try {
                    val f = NativeInputEvent::class.java.getDeclaredField("reserved")
                    f.isAccessible = true
                    f.setShort(e, 0x01.toShort())
                } catch (x: Exception) {
                    logger.error(x) { "Failed to consume native key event" }
                }

                executor(a)
            }
        }

        fun release() {
            isPressed = false
        }
    }

    companion object: KLogging()

}